### PR TITLE
doc (collections): Add contribtution guide, some principles and extend README

### PR DESCRIPTION
**Pull request by LionC**
This adds a contribution guide with a checklist for common things I have been seeing in recent PRs to the module. It also
adds a small philosophy section that tries to summarize the current ideas behind the module and which might answer common
questions.

It also adds a bit of info to the README and updates missing functions.

### CODE REVIEW

File: collections/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+in this folder, read through issues and PRs regarding the `collections` module or visit the
+[#deno-std channel on our Discord](https://discord.gg/Ub9bRxHv) and ask.
+
+## Philosophy
+
+Currently, those are the general ideas behind how the module is implemented:
+
+- Provide a clear, specific toolbox and vocabulary for common tasks with `Array`s and `Record`s
+- Optimize each specific tasks for maximum runtime and memory efficiency
+- Offer only pure functions, never mutate given arguments, always return a new value
```
Comment (kt3k) [inline collections/CONTRIBUTING.md:13]: I feel we are open to including mutating functions when it makes sense. For example, deepAssign suggestion seems worth the consideration in my view.

**Review by kt3k — COMMENTED** [review]
@LionC Thank you for the contribution. Left one comment. Also this needs formatting by `deno fmt`

File: collections/CONTRIBUTING.md (hunk 1 above)
Comment (LionC) [reply collections/CONTRIBUTING.md]: Yes I agree. Maybe change to "functions are pure by default, side effect need to be clearly marked and putt into <folder>"?

Comment (kt3k) [reply collections/CONTRIBUTING.md]: > Functions are expected to be pure by default, side effect (like mutation of arguments) need to be clearly documented and the function put into someFolderNameWeNeedToFigureOut

Sounds good to me.

Comment (kt3k) [reply collections/CONTRIBUTING.md]: The folder name might be `collections/mutate`, `collections/mut`? I'm even ok with no special folder for mutating functions if it's documented well.

Comment (LionC) [reply collections/CONTRIBUTING.md]: I will change this today 👌

**Comment by LionC** [comment]
Done @kt3k

**Review by kt3k — APPROVED** [review]
@LionC LGTM. Thanks! Nice work!

**Pull request merged**